File: Algo/Leetcode/040CombinationSumII.cpp
// dfs
class Solution {
public:
    vector<vector<int>> res;
    vector<int> r;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        int n = candidates.size();
        if(n==0) return res;
        sort(candidates.begin(), candidates.end());  //sort first
        dfs(candidates, target, 0, 0);
        return res;
    }

    void dfs(vector<int>& candidates, int target, int sum, int start){
        if(sum == target){
            res.push_back(r);
            return;
        }

        if(sum > target){
            return;
        }

        for(int i=start; i<candidates.size(); i++){
            if(i>start && candidates[i]==candidates[i-1])  // in case repetitive selection
                continue;
            r.push_back(candidates[i]);
            sum += candidates[i];
            dfs(candidates, target, sum, i+1);
            r.pop_back();
            sum -= candidates[i];
        }
    }
};

File: Algo/Leetcode/040CombinationSumII.py
class Solution(object):

    def __init__(self):
        self.res = []
        self.path = []
        self.sum = 0

    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates = sorted(candidates)
        n = len(candidates)
        self.dfs(candidates, 0, n, target)

        return self.res

    def dfs(self, candidates, i, n, target):
        if self.sum == target:
            self.res.append(copy.deepcopy(self.path))
            return

        if i >= n or self.sum > target:
            return

        for j in range(i, n):
            if j == i or candidates[j] != candidates[j - 1]:  # DO NOT PICK the repeat one
                self.sum += candidates[j]
                self.path.append(candidates[j])
                self.dfs(candidates, j + 1, n, target)
                self.path.pop()
                self.sum -= candidates[j]